enclave/sealing/sgx/sealer_gcm_aes128.c

```diff
@@ -0,0 +1,296 @@
+                break;
+
+            case OE_SEAL_SETTING_SGX_XFRMMASK:
+                header->keyrequest.attribute_mask.xfrm = settings[i].value.q;
+                break;
+
+            case OE_SEAL_SETTING_SGX_MISCMASK:
+                header->keyrequest.misc_attribute_mask = settings[i].value.d;
+                break;
+
+            case OE_SEAL_SETTING_SGX_CONFIGSVN:
+                header->keyrequest.config_svn = settings[i].value.w;
+                break;
+
+            case OE_SEAL_SETTING_ADDITIONAL_CONTEXT:
+                // Custom OE_SEAL_SETTING_ADDITIONAL_CONTEXT not supported
+
+            case OE_SEAL_SETTING_SGX_CET_ATTRIBUTES_MASK:
+                // OE_SEAL_SETTING_SGX_CET_ATTRIBUTES_MASK not supported
+
+            default:
+                OE_RAISE(OE_UNSUPPORTED);
+        }
+
+    OE_CHECK(oe_get_seal_key(
+        (uint8_t*)&header->keyrequest,
+        sizeof(header->keyrequest),
+        &key,
+        &size));
+
+    OE_STATIC_ASSERT(sizeof(header->tag) >= 16);
+    OE_CHECK(oe_aes_gcm_encrypt(
+        key,
+        size,
+        header->iv,
+        sizeof(header->iv),
+        additional_data,
+        additional_data_size,
+        plaintext,
+        plaintext_size,
+        (uint8_t*)(header + 1),
+        header->tag));
+
+done:
+    oe_free_key(key, size, NULL, 0);
+
+    if (result != OE_OK)
+    {
+        oe_free(*blob);
+        *blob = NULL;
+    }
+
+    return result;
+}
+
+static oe_result_t _unseal(
+    const uint8_t* blob,
+    size_t blob_size,
+    const uint8_t* additional_data,
+    size_t additional_data_size,
+    uint8_t** plaintext,
+    size_t* plaintext_size)
+{
+    oe_result_t result = OE_OK;
+    struct _sealed_blob_header* header = (struct _sealed_blob_header*)blob;
+    uint8_t* key = NULL;
+    size_t key_size = 0;
+
+    if (blob_size < sizeof(*header))
+        return OE_INVALID_PARAMETER;
+
+    if (header->ciphertext_size != blob_size - sizeof(*header) ||
+        header->total_size != header->ciphertext_size + additional_data_size)
+        return OE_CRYPTO_ERROR;
+
+    *plaintext_size = header->ciphertext_size;
+    *plaintext = (uint8_t*)oe_calloc(1, *plaintext_size);
+    if (*plaintext == NULL)
+        return OE_OUT_OF_MEMORY;
+
+    OE_CHECK(oe_get_seal_key(
+        (uint8_t*)&header->keyrequest,
+        sizeof(header->keyrequest),
+        &key,
+        &key_size));
+
+    OE_CHECK(oe_aes_gcm_decrypt(
+        key,
+        key_size,
+        header->iv,
+        sizeof(header->iv),
+        additional_data,
+        additional_data_size,
+        (uint8_t*)(header + 1),
+        header->ciphertext_size,
+        *plaintext,
+        header->tag));
+
+done:
+    oe_free_key(key, key_size, NULL, 0);
+
+    if (result != OE_OK)
+    {
+        oe_free(*plaintext);
+        *plaintext = NULL;
+    }
+
+    return result;
+}
+
+// The follwoing won't work because this seal plugin will be discarded by ld as
```

Comment:
To elaborate: the issue is that we distribute the OE SDK libraries as archive (.a) files and the constructor attribute apparently does not result in a `DT_INIT` entry in the archive, so when the linker (ld) comes along, the constructor is just treated as another function, and can be eliminated if unreferenced. It looks like the constructor is correctly maintained if built directly into a binary as in our test cases. 

While we could solve this by making it a required into into the linker command line, that would be a burden on the end-user and undesirable as a result. 

For OE SDK choice of default plugins, their registration can probably be consolidated into the `_handle_init_enclave()` function which is required to run on `oe_create_enclave()`. 

@anakrish if he has any additional opinions on this.

Reply:
Putting it into `_handle_init_enclave()` would link sealing code into every enclave, regardless of whether it's used or not.

If *OE SDK choice of default* is the way to go, I could reference it in `oe_seal()`/`oe_unseal`'s implementation so it would be linked in only if the enclave uses sealing. However, that means the default plug-in will always be linked in even if the user specifies and only uses his/her own plug-in.

Yet another solution is to make the plug-in "library" an object (.o) file rather than an archive (.a). Then ld will always keep the constructor. This is currently my tendency.

Reply:
Yeah, I agree that a .obj file is probably preferable if you want the sealing plugin to be a separate target that needs to be included (i.e. not included as part of liboeenclave.a). I think it's okay for liboeenclave to provide a default sealing implementation if the reason we're providing sealing APIs is to free devs from using the existing `oe_get_seal_key` APIs and rolling their own use of it. I would reference how the attestation plugins handle this; I think they end up always linking in the default implementations, so they might be a reasonable reference point. 

One additional note: the specific choice of releasing the component as an object (.obj) may also be a temporary constraint. There is support for loading enclaves consisting of multiple shared object (.so) files coming up, and sources with the constructor attribute built directly into an .so binary should also work (the `DT_INIT` info is built into the .so and the dynamic binding code in the enclave finds and invokes all global initializers).